File: client/android/app/src/main/java/app/mobiling/client/access/MobilingAppShell.kt
package app.mobiling.client.access

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import app.mobiling.client.attachment.AttachmentFeatureBridge
import app.mobiling.client.auth.AccessAuthFeatureBridge
import app.mobiling.client.catalog.CatalogFeatureBridge
import app.mobiling.client.contract.auth.session.AccessAuthSessionPayload
import app.mobiling.client.cart.CartFeatureBridge
import app.mobiling.client.RetailKind
import app.mobiling.client.dashboard.DashboardMobileShell
import app.mobiling.client.data.navigation.shell.NavigationShellGateway
import app.mobiling.client.message.MessageFeatureBridge
import app.mobiling.client.notification.NotificationFeatureBridge
import app.mobiling.client.support.SupportFeatureBridge
import app.mobiling.client.data.order.OrderGateway
import app.mobiling.client.data.product.ProductGateway
import app.mobiling.client.data.project.ProjectGateway
import app.mobiling.client.data.retail.placement.RetailPlacementGateway
import app.mobiling.client.data.vendor.payout.VendorPayoutGateway
import app.mobiling.client.data.vendor.profile.VendorProfileGateway
import app.mobiling.client.data.vendor.statement.VendorStatementGateway
import app.mobiling.client.data.vendor.summary.VendorSummaryGateway
import app.mobiling.client.data.vendor.transaction.VendorTransactionGateway
import app.mobiling.client.data.wallet.WalletGateway
import kotlinx.coroutines.launch

@Composable
fun MobilingAppShell(
    accessAuthFeatureBridge: AccessAuthFeatureBridge? = null,
    cartFeatureBridge: CartFeatureBridge? = null,
    attachmentFeatureBridge: AttachmentFeatureBridge? = null,
    catalogFeatureBridge: CatalogFeatureBridge? = null,
    navigationShellGateway: NavigationShellGateway? = null,
    messageFeatureBridge: MessageFeatureBridge? = null,
    notificationFeatureBridge: NotificationFeatureBridge? = null,
    supportFeatureBridge: SupportFeatureBridge? = null,
    productGateway: ProductGateway? = null,
    retailPlacementGateway: RetailPlacementGateway? = null,
    orderGateway: OrderGateway? = null,
    projectGateway: ProjectGateway? = null,
    vendorProfileGateway: VendorProfileGateway? = null,
    vendorSummaryGateway: VendorSummaryGateway? = null,
    vendorStatementGateway: VendorStatementGateway? = null,
    vendorPayoutGateway: VendorPayoutGateway? = null,
    vendorTransactionGateway: VendorTransactionGateway? = null,
    walletGateway: WalletGateway? = null,
    initialRoute: String = "dashboard",
    publicInitialRoute: String = "home",
    catalogEnabled: Boolean = true,
    availableRetailKinds: List<RetailKind> = RetailKind.entries,
    navigationLabelResolver: (route: String?, key: String, backendLabel: String) -> String = { _, _, label -> label },
    onAuthenticated: suspend () -> Unit = {},
    onBeforeSignOut: suspend () -> Unit = {},
    authenticatedContent: (@Composable (vendorId: String?, onSignOut: () -> Unit) -> Unit)? = null,
) {
    var currentScreen by remember(publicInitialRoute) {
        mutableStateOf(if (publicInitialRoute == "sign-in") AccessScreen.SignIn else AccessScreen.Welcome)
    }
    var activeVendorId by remember { mutableStateOf<String?>(null) }
    var activeUserUuid by remember { mutableStateOf<String?>(null) }
    val coroutineScope = rememberCoroutineScope()

    fun applyAccessSession(payload: AccessAuthSessionPayload) {
        activeVendorId = payload.vendorId
        activeUserUuid = payload.userUuid
        currentScreen = payload.toAccessScreen()
        if (currentScreen == AccessScreen.Dashboard) {
            coroutineScope.launch { onAuthenticated() }
        }
    }

    fun clearAccessSession() {
        coroutineScope.launch {
            try {
                onBeforeSignOut()
            } catch (_: Exception) {
            }

            try {
                accessAuthFeatureBridge?.logout()
            } catch (_: Exception) {
            }

            activeVendorId = null
            activeUserUuid = null
            currentScreen = if (publicInitialRoute == "sign-in") AccessScreen.SignIn else AccessScreen.Welcome
        }
    }

    LaunchedEffect(accessAuthFeatureBridge) {
        val payload = try {
            accessAuthFeatureBridge?.restore()
        } catch (_: Exception) {
            null
        }

        if (payload != null) {
            applyAccessSession(payload)
        }
    }

    Surface(Modifier.fillMaxSize()) {
        Box(Modifier.fillMaxSize()) {
            when (currentScreen) {
            AccessScreen.Dashboard -> {
                if (authenticatedContent != null) {
                    authenticatedContent(activeVendorId) { clearAccessSession() }
                } else {
                    DashboardMobileShell(
                        navigationShellGateway = navigationShellGateway,
                        messageFeatureBridge = messageFeatureBridge,
                        notificationFeatureBridge = notificationFeatureBridge,
                        supportFeatureBridge = supportFeatureBridge,
                        productGateway = productGateway,
                        retailPlacementGateway = retailPlacementGateway,
                        orderGateway = orderGateway,
                        projectGateway = projectGateway,
                        cartFeatureBridge = cartFeatureBridge,
                        catalogFeatureBridge = catalogFeatureBridge,
                        attachmentFeatureBridge = attachmentFeatureBridge,
                        vendorId = activeVendorId,
                        userUuid = activeUserUuid,
                        vendorProfileGateway = vendorProfileGateway,
                        vendorSummaryGateway = vendorSummaryGateway,
                        vendorStatementGateway = vendorStatementGateway,
                        vendorPayoutGateway = vendorPayoutGateway,
                        vendorTransactionGateway = vendorTransactionGateway,
                        walletGateway = walletGateway,
                        initialRoute = initialRoute,
                        catalogEnabled = catalogEnabled,
                        availableRetailKinds = availableRetailKinds,
                        navigationLabelResolver = navigationLabelResolver,
                        onSignOut = { clearAccessSession() },
                    )
                }
            }

            AccessScreen.Welcome -> AccessWelcomeScreen(
                initialRoute = publicInitialRoute,
                catalogFeatureBridge = catalogFeatureBridge,
                cartFeatureBridge = cartFeatureBridge,
                onSignIn = { currentScreen = AccessScreen.SignIn },
                onCreateAccess = { currentScreen = AccessScreen.Register },
            )

            AccessScreen.SignIn -> SignInScreen(
                onBack = { currentScreen = AccessScreen.Welcome },
                onCreateAccess = { currentScreen = AccessScreen.Register },
                onRecoverAccess = { currentScreen = AccessScreen.RecoveryRequest },
                onStartAccess = { request -> accessAuthFeatureBridge?.start(request) },
                onAccessSession = { payload -> applyAccessSession(payload) },
            )

            AccessScreen.Register -> RegisterAccessScreen(
                onBack = { currentScreen = AccessScreen.Welcome },
                onSignIn = { currentScreen = AccessScreen.SignIn },
                onRegisterAccess = { request -> accessAuthFeatureBridge?.register(request) },
                onAccessSession = { payload -> applyAccessSession(payload) },
            )

            AccessScreen.VerificationRequired -> VerificationRequiredScreen(
                onBack = { currentScreen = AccessScreen.SignIn },
                onUseDifferentAccess = { clearAccessSession() },
            )

            AccessScreen.SecondFactorRequired -> SecondFactorRequiredScreen(
                onBack = { currentScreen = AccessScreen.SignIn },
                onUseDifferentAccess = { clearAccessSession() },
            )

            AccessScreen.RecoveryRequest -> RecoveryRequestScreen(
                onBack = { currentScreen = AccessScreen.SignIn },
                onHaveRecoveryCode = { currentScreen = AccessScreen.RecoveryReset },
                onRequestRecovery = { request -> accessAuthFeatureBridge?.requestRecovery(request) },
                onAccessSession = { payload -> applyAccessSession(payload) },
            )

            AccessScreen.RecoveryReset -> RecoveryResetScreen(
                onBack = { currentScreen = AccessScreen.RecoveryRequest },
                onRequestRecovery = { currentScreen = AccessScreen.RecoveryRequest },
                onResetRecovery = { request -> accessAuthFeatureBridge?.resetRecovery(request) },
                onAccessSession = { payload -> applyAccessSession(payload) },
            )
            }
        }
    }
}

File: client/ios/App/Access/MobilingAppShell.swift
import SwiftUI

public struct MobilingAppShell: View {
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var currentScreen: AccessScreen
    @State private var activeVendorId: String?
    @State private var launchSplashMounted: Bool = true
    @State private var launchSplashVisible: Bool = true
    @State private var launchSplashDismissScheduled: Bool = false
    private let authFeatureBridge: AuthFeatureBridge?
    private let attachmentFeatureBridge: AttachmentFeatureBridge?
    private let cartFeatureBridge: CartFeatureBridge?
    private let catalogFeatureBridge: CatalogFeatureBridge?
    private let navigationShellGateway: NavigationShellGateway?
    private let messageFeatureBridge: MessageFeatureBridge?
    private let notificationFeatureBridge: NotificationFeatureBridge?
    private let vendorProfileGateway: VendorProfileGateway?
    private let vendorSummaryGateway: VendorSummaryGateway?
    private let vendorStatementGateway: VendorStatementGateway?
    private let vendorPayoutGateway: VendorPayoutGateway?
    private let vendorTransactionGateway: VendorTransactionGateway?
    private let vendorCrudGateway: VendorCrudGateway?
    private let retailPlacementGateway: RetailPlacementGateway?
    private let walletGateway: WalletGateway?
    private let initialRoute: String
    private let publicInitialRoute: String
    private let catalogEnabled: Bool
    private let availableRetailKinds: [RetailKind]
    private let navigationLabelResolver: (String?, String, String) -> String
    private let onAuthenticated: () async -> Void
    private let onBeforeSignOut: () async -> Void

    public init(
        authFeatureBridge: AuthFeatureBridge? = nil,
        attachmentFeatureBridge: AttachmentFeatureBridge? = nil,
        cartFeatureBridge: CartFeatureBridge? = nil,
        catalogFeatureBridge: CatalogFeatureBridge? = nil,
        navigationShellGateway: NavigationShellGateway? = nil,
        messageFeatureBridge: MessageFeatureBridge? = nil,
        notificationFeatureBridge: NotificationFeatureBridge? = nil,
        vendorProfileGateway: VendorProfileGateway? = nil,
        vendorSummaryGateway: VendorSummaryGateway? = nil,
        vendorStatementGateway: VendorStatementGateway? = nil,
        vendorPayoutGateway: VendorPayoutGateway? = nil,
        vendorTransactionGateway: VendorTransactionGateway? = nil,
        vendorCrudGateway: VendorCrudGateway? = nil,
        retailPlacementGateway: RetailPlacementGateway? = nil,
        walletGateway: WalletGateway? = nil,
        initialRoute: String = "dashboard",
        publicInitialRoute: String = "home",
        catalogEnabled: Bool = true,
        availableRetailKinds: [RetailKind] = RetailKind.allCases,
        navigationLabelResolver: @escaping (String?, String, String) -> String = { _, _, label in label },
        onAuthenticated: @escaping () async -> Void = {},
        onBeforeSignOut: @escaping () async -> Void = {}
    ) {
        self.authFeatureBridge = authFeatureBridge
        self.attachmentFeatureBridge = attachmentFeatureBridge
        self.cartFeatureBridge = cartFeatureBridge
        self.catalogFeatureBridge = catalogFeatureBridge
        self.navigationShellGateway = navigationShellGateway
        self.messageFeatureBridge = messageFeatureBridge
        self.notificationFeatureBridge = notificationFeatureBridge
        self.vendorProfileGateway = vendorProfileGateway
        self.vendorSummaryGateway = vendorSummaryGateway
        self.vendorStatementGateway = vendorStatementGateway
        self.vendorPayoutGateway = vendorPayoutGateway
        self.vendorTransactionGateway = vendorTransactionGateway
        self.vendorCrudGateway = vendorCrudGateway
        self.retailPlacementGateway = retailPlacementGateway
        self.walletGateway = walletGateway
        self.initialRoute = initialRoute
        self.publicInitialRoute = publicInitialRoute
        _currentScreen = State(initialValue: publicInitialRoute == "sign-in" ? .signIn : .welcome)
        self.catalogEnabled = catalogEnabled
        self.availableRetailKinds = availableRetailKinds
        self.navigationLabelResolver = navigationLabelResolver
        self.onAuthenticated = onAuthenticated
        self.onBeforeSignOut = onBeforeSignOut
    }

    public var body: some View {
        ZStack {
            Group {
                switch currentScreen {
            case .dashboard:
                MobileDashboardShellView(
                    navigationShellGateway: navigationShellGateway,
                    messageFeatureBridge: messageFeatureBridge,
                    notificationFeatureBridge: notificationFeatureBridge,
                    attachmentFeatureBridge: attachmentFeatureBridge,
                    cartFeatureBridge: cartFeatureBridge,
                    catalogFeatureBridge: catalogFeatureBridge,
                    vendorId: activeVendorId,
                    vendorProfileGateway: vendorProfileGateway,
                    vendorSummaryGateway: vendorSummaryGateway,
                    vendorStatementGateway: vendorStatementGateway,
                    vendorPayoutGateway: vendorPayoutGateway,
                    vendorTransactionGateway: vendorTransactionGateway,
                    vendorCrudGateway: vendorCrudGateway,
                    retailPlacementGateway: retailPlacementGateway,
                    walletGateway: walletGateway,
                    initialRoute: initialRoute,
                    catalogEnabled: catalogEnabled,
                    availableRetailKinds: availableRetailKinds,
                    navigationLabelResolver: navigationLabelResolver,
                    onSignOut: { clearAccessSession() }
                )
            case .welcome:
                AccessWelcomeView(
                    initialRoute: publicInitialRoute,
                    catalogFeatureBridge: catalogFeatureBridge,
                    cartFeatureBridge: cartFeatureBridge,
                    onSignIn: { currentScreen = .signIn },
                    onCreateAccess: { currentScreen = .register }
                )
            case .signIn:
                SignInView(
                    onBack: { currentScreen = .welcome },
                    onCreateAccess: { currentScreen = .register },
                    onRecoverAccess: { currentScreen = .recoveryRequest },
                    onStartAccess: { request in
                        guard let authFeatureBridge else {
                            return nil
                        }
                        return try await authFeatureBridge.start(request: request)
                    },
                    onAccessSession: { payload in applyAccessSession(payload) }
                )
            case .register:
                RegisterAccessView(
                    onBack: { currentScreen = .welcome },
                    onSignIn: { currentScreen = .signIn },
                    onRegisterAccess: { request in
                        guard let authFeatureBridge else {
                            return nil
                        }
                        return try await authFeatureBridge.register(request: request)
                    },
                    onAccessSession: { payload in applyAccessSession(payload) }
                )
            case .verificationRequired:
                VerificationRequiredView(
                    onBack: { currentScreen = .signIn },
                    onUseDifferentAccess: { clearAccessSession() }
                )
            case .secondFactorRequired:
                SecondFactorRequiredView(
                    onBack: { currentScreen = .signIn },
                    onUseDifferentAccess: { clearAccessSession() }
                )
            case .recoveryRequest:
                RecoveryRequestView(
                    onBack: { currentScreen = .signIn },
                    onHaveRecoveryCode: { currentScreen = .recoveryReset },
                    onRequestRecovery: { request in
                        guard let authFeatureBridge else {
                            return nil
                        }
                        return try await authFeatureBridge.requestRecovery(request: request)
                    },
                    onAccessSession: { payload in applyAccessSession(payload) }
                )
            case .recoveryReset:
                RecoveryResetView(
                    onBack: { currentScreen = .recoveryRequest },
                    onRequestRecovery: { currentScreen = .recoveryRequest },
                    onResetRecovery: { request in
                        guard let authFeatureBridge else {
                            return nil
                        }
                        return try await authFeatureBridge.resetRecovery(request: request)
                    },
                    onAccessSession: { payload in applyAccessSession(payload) }
                )
                }
            }

            if launchSplashMounted {
                OneTaskerLaunchSplashView(isVisible: launchSplashVisible)
                    .zIndex(10)
            }
        }
        .environment(\.mobileMessageComposer, MobileDesignDefaults.messageComposer)
        .onAppear { scheduleLaunchSplashDismiss() }
        .task {
            guard let authFeatureBridge else {
                return
            }

            do {
                applyAccessSession(try await authFeatureBridge.restore())
            } catch {
            }
        }
    }

    private func applyAccessSession(_ payload: AuthSessionPayload) {
        activeVendorId = payload.vendorId
        currentScreen = payload.toAccessScreen()
        if currentScreen == .dashboard {
            Task { await onAuthenticated() }
        }
    }

    private func clearAccessSession() {
        Task {
            await onBeforeSignOut()

            do {
                try await authFeatureBridge?.logout()
            } catch {
            }

            activeVendorId = nil
            currentScreen = publicInitialRoute == "sign-in" ? .signIn : .welcome
        }
    }

    private func scheduleLaunchSplashDismiss() {
        guard !launchSplashDismissScheduled else {
            return
        }

        launchSplashDismissScheduled = true
        Task { @MainActor in
            let holdNanoseconds: UInt64 = reduceMotion ? 520_000_000 : 7_200_000_000
            let fadeNanoseconds: UInt64 = reduceMotion ? 120_000_000 : 320_000_000

            try? await Task.sleep(nanoseconds: holdNanoseconds)
            withAnimation(.easeOut(duration: reduceMotion ? 0.08 : 0.28)) {
                launchSplashVisible = false
            }

            try? await Task.sleep(nanoseconds: fadeNanoseconds)
            launchSplashMounted = false
        }
    }
}

